src/main/java/org/myrobotlab/service/NeoPixel.java

```diff
@@ -26,501 +26,829 @@
-  public boolean off = false;
-
-  public static transient final int NEOPIXEL_ANIMATION_NO_ANIMATION = 0;
-  public static transient final int NEOPIXEL_ANIMATION_STOP = 1;
-  public static transient final int NEOPIXEL_ANIMATION_COLOR_WIPE = 2;
-  public static transient final int NEOPIXEL_ANIMATION_LARSON_SCANNER = 3;
-  public static transient final int NEOPIXEL_ANIMATION_THEATER_CHASE = 4;
-  public static transient final int NEOPIXEL_ANIMATION_THEATER_CHASE_RAINBOW = 5;
-  public static transient final int NEOPIXEL_ANIMATION_RAINBOW = 6;
-  public static transient final int NEOPIXEL_ANIMATION_RAINBOW_CYCLE = 7;
-  public static transient final int NEOPIXEL_ANIMATION_FLASH_RANDOM = 8;
-  public static transient final int NEOPIXEL_ANIMATION_IRONMAN = 9;
-
-  public List<String> animations = Arrays.asList("No animation", "Stop", "Color Wipe", "Larson Scanner", "Theater Chase", "Theater Chase Rainbow", "Rainbow", "Rainbow Cycle",
-      "Flash Random", "Ironman");
-  public transient String animation = "No animation";
-  public transient boolean[] animationSetting = { false, false }; // red,
-  // green,
-  // blue,
-  // speed
-  public transient boolean animationSettingColor = false;
-  public transient boolean animationSettingSpeed = false;
-  transient HashMap<Integer, boolean[]> animationSettings = new HashMap<Integer, boolean[]>();
+  protected Integer pixelCount = null;
+
+  /**
+   * RGB or RGBW supported 3 RGB 4 RGBW
+   */
+  protected int pixelDepth = 3;
+  /**
+   * current selected red value
+   */
+  
+  protected int red = 0;
+
+  /**
+   * currently selected animation
+   */
+  protected String currentAnimation;
+
+  /**
+   * speed of an animation in fps
+   */
+  protected int speedFps = 10;
+  
+  /**
+   * map between speed and wait time in ms for those animations that
+   * need it (onboard)
+   */
+  protected MapperLinear fpsToWaitMs = new MapperLinear(1, 30, 1000, 30);
 
   public NeoPixel(String n, String id) {
     super(n, id);
-    animationSettings.put(NEOPIXEL_ANIMATION_STOP, new boolean[] { false, false });
-    animationSettings.put(NEOPIXEL_ANIMATION_COLOR_WIPE, new boolean[] { true, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_LARSON_SCANNER, new boolean[] { true, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_THEATER_CHASE, new boolean[] { true, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_THEATER_CHASE_RAINBOW, new boolean[] { false, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_RAINBOW, new boolean[] { false, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_RAINBOW_CYCLE, new boolean[] { false, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_FLASH_RANDOM, new boolean[] { true, true });
-    animationSettings.put(NEOPIXEL_ANIMATION_IRONMAN, new boolean[] { true, true });
-    subscribeToRuntime("registered");
-  }
-
-  public void onRegistered(Registration s) {
-    refreshControllers();
-    broadcastState();
+    animationRunner = new AnimationRunner();

```

Comment:
We shouldn't be creating this thread when the service is created.  This should go into the startService lifecycle , not here.   I'm not sure we have a need for the animation runner.  animations run on the microcontroller in mrlcomm.  if users want to create an animation, they would likely write their own python code to script that up.

Reply:
it has to be in the constructor - its "final" - its not a thread - its a runnable that manages its own thread ..
I prefer to design like this - its clean - you create it once - it manages its own thread, less likely to create "multiple" and other thread messes